Logic/igtlioConnector.cxx

```diff
@@ -341,209 +310,220 @@ void igtlioConnector::RequestPushOutgoingMessages()
+  // This may need to be parallelized so that other socket aren't waiting on timeouts
+  int r = client.Socket->Receive(headerMsg->GetPackPointer(), headerMsg->GetPackSize());
 
-      if (r != headerMsg->GetPackSize())
-      {
-        vtkDebugMacro("ignoring header, breaking. received=" << r);
-        break;
-      }
+  vtkDebugMacro("Received header of size: " << headerMsg->GetPackSize());
 
-      // Deserialize the header
-      headerMsg->Unpack();
+  if (r != headerMsg->GetPackSize())
+  {
+    vtkDebugMacro("ignoring header, breaking. received=" << r);
+    this->RemoveClient(client.ID);
+    return false;
+  }
 
-      //----------------------------------------------------------------
-      // Check Device Name
-      // Nov 16, 2010: Currently the following code only checks
-      // if the device name is defined in the message.
-      const char* devName = headerMsg->GetDeviceName();
-      if (devName[0] == '\0')
-      {
-        /// Dec 7, 2010: Removing the following code, since message without
-        /// device name should be handled in the MRML scene as well.
-        //// If no device name is defined, skip processing the message.
-        //this->Skip(headerMsg->GetBodySizeToRead());
-        //continue; //  while (!this->ServerStopFlag)
-      }
-      //----------------------------------------------------------------
-      // If device name is restricted
-      else if (this->RestrictDeviceName)
-      {
-        // Check if the node has already been registered.
-          //TODO: Cannot call GetDevice in Thread!!!!
-        igtlioDeviceKeyType key = igtlioDeviceKeyType::CreateDeviceKey(headerMsg);
-        int registered = this->GetDevice(key).GetPointer() != NULL;
-        if (registered == 0)
-        {
-          this->Skip(headerMsg->GetBodySizeToRead(), client);
-          continue; //  while (!this->ServerStopFlag)
-        }
-      }
+  // Deserialize the header
+  headerMsg->Unpack();
 
-      vtkDebugMacro("completed read header : " << headerMsg->GetDeviceName() << " body size to read: " << headerMsg->GetBodySizeToRead());
+  //----------------------------------------------------------------
+  // Check Device Name
+  // Nov 16, 2010: Currently the following code only checks
+  // if the device name is defined in the message.
+  const char* devName = headerMsg->GetDeviceName();
+  if (devName[0] == '\0')
+  {
+    /// Dec 7, 2010: Removing the following code, since message without
+    /// device name should be handled in the MRML scene as well.
+    //// If no device name is defined, skip processing the message.
+    //this->Skip(headerMsg->GetBodySizeToRead());
+    //continue; //  while (!this->ServerStopFlag)
+  }
+  //----------------------------------------------------------------
+  // If device name is restricted
+  else if (this->RestrictDeviceName)
+  {
+    // Check if the node has already been registered.
+      //TODO: Cannot call GetDevice in Thread!!!!
```

Comment:
If it is important to fix this then consider adding an issue to the issue tracker to make sure we don't forget about it (unless you plan to fix this right away).

Reply:
I didn't add this comment, It's been there for a while, but I can make an issue for it.

Reply:
OK, thanks. Maybe spend a very little time investigating how severe this issue actually is (or if it is still an issue at all) and write up the result in an issue if there is no trivial solution.

Reply:
If it's a muted lock issue we could discuss a recursive mutex

Reply:
GetDevice(igtlioDeviceKeyType key) iterates through the list of devices without a mutex lock, which is likely what is being reference by the content.

Other threads could add/remove elements and cause a crash.

Same thing with AddDevice and RemoveDevice

Reply:
Ok those definitely need to be made thread safe.

Thanks so much for finishing this task for me. Very much appreciated.

Reply:
Related to the recursive mutex, there isn't a read/write mutex in VTK, is there?

Reply:
I don't think there is a read/write mutex in VTK.